rclcpp/include/rclcpp/executors.hpp

```diff
@@ -79,10 +79,11 @@ using rclcpp::executors::SingleThreadedExecutor;
  *   If the time spent inside the blocking loop exceeds this timeout, return a `TIMEOUT` return code.
  * \return The return code, one of `SUCCESS`, `INTERRUPTED`, or `TIMEOUT`.
  */
-template<typename FutureT, typename TimeRepT = int64_t, typename TimeT = std::milli>
+template<typename FutureT, typename Executor, typename TimeRepT = int64_t,
+  typename TimeT = std::milli>
 rclcpp::FutureReturnCode
 spin_node_until_future_complete(
-  rclcpp::Executor & executor,
+  Executor & executor,
```

Comment:
Why does this need to be a template argument? Are we actually passing an executor here that doesn't use `rclcpp::Executor` as a base?

Reply:
Yes, I don't derive from rclcpp:: Executor

Reply:
Is that something we actually need to support? I'd like to understand the use case here a little bit more. I'm suspicious of doing this. It prevents us from clearly defining the API of an executor and instead requires us to use these template functions that kind of duck-type object that was passed in.

Instead, I'd prefer to settle on changes to the base class to make it easier to inherit from, rather than make that optional.

Reply:
For example, if the data in the `rclcpp::Executor` base class is not desirable, we should make a new base for that called like `rclcpp::ExecutorInterface` which defines the required API.

Reply:
+1 on using an executor base class

Reply:
makes sense, we want to use base class to avoid taking this path if possible.

Reply:
I agree, a base class is more desirable. Big question though, can we make this happen until monday ?

Reply:
As to why I need / want this change: 

rclcpp::Executor currently does stuff in the constructor that is not desirable for me, like registering callbacks etc.
I also hat name collisions with my variables and the ones already present.

Some methods of rclcpp::Executor are not virtual, therefore I will get a wrong behavior, if someone would call them on my Executor.
E.g. : 
https://github.com/ros2/rclcpp/blob/90e451154cc23c4c231ca55dc103358f38cfd999/rclcpp/include/rclcpp/executor.hpp#L247

https://github.com/ros2/rclcpp/blob/90e451154cc23c4c231ca55dc103358f38cfd999/rclcpp/include/rclcpp/executor.hpp#L276

https://github.com/ros2/rclcpp/blob/90e451154cc23c4c231ca55dc103358f38cfd999/rclcpp/include/rclcpp/executor.hpp#L281

https://github.com/ros2/rclcpp/blob/90e451154cc23c4c231ca55dc103358f38cfd999/rclcpp/include/rclcpp/executor.hpp#L311

https://github.com/ros2/rclcpp/blob/90e451154cc23c4c231ca55dc103358f38cfd999/rclcpp/include/rclcpp/executor.hpp#L318

https://github.com/ros2/rclcpp/blob/90e451154cc23c4c231ca55dc103358f38cfd999/rclcpp/include/rclcpp/executor.hpp#L424

https://github.com/ros2/rclcpp/blob/90e451154cc23c4c231ca55dc103358f38cfd999/rclcpp/include/rclcpp/executor.hpp#L433

To me the proposed changes looked the least intrusive.

Reply:
> I agree, a base class is more desirable. Big question though, can we make this happen until monday ?

The thing is that if we merge this just because Monday is coming up, then we have to come back and undo what we merged through a deprecation cycle, which is very undesirable.

> https://github.com/ros2/rclcpp/blob/90e451154cc23c4c231ca55dc103358f38cfd999/rclcpp/include/rclcpp/executor.hpp#L247

This one is a template function, so that cannot be virtual and you shouldn't need to override that, only the function that it calls.

For the other functions, I'd prefer to see those get marked `virtual` than to make the executor base class not a requirement.

Reply:
> This one is a template function, so that cannot be virtual and you shouldn't need to override that, only the function that it calls.

I am aware of that, but the function it calls is also not virtual.

Actually, why is this a template function in the first place ?

Reply:
> Actually, why is this a template function in the first place ?

It's the easiest way to handle a wide array of input chrono types for the timeout, at least as far as I know. Perhaps there's a better way to do that now, but at the time I believe that's why it was made that way.